api/v1/address.go

```diff
@@ -1,55 +1,133 @@
 package api
 
 import (
+	"encoding/json"
 	"net/http"
 
+	"github.com/data-preservation-programs/spade-tenant/core"
+	"github.com/data-preservation-programs/spade-tenant/db"
 	"github.com/labstack/echo/v4"
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
 )
 
-type Addresses []Address
-type AddressesMutable []AddressMutable
+type Addresses []db.Address
+type Address db.Address
 
-type Address struct {
-	AddressMutable
-	ActorID *uint `json:"actor_id,omitempty"` // TODO :swagger docs should be null not 0
+type AddressResponse struct {
+	AddressRobust    string `json:"address_robust"`
+	AddressActorID   uint   `json:"actor_id"`
+	AddressIsSigning bool   `json:"is_signing"`
 }
 
-type AddressMutable struct {
-	Address   string `json:"address"`
-	IsSigning bool   `json:"is_signing"` // true - active dealmaking from this address, false - still counts as an associated wallet
+var service *core.SpdTenantSvc
+
+func ConfigureAddressesRouter(e *echo.Group, tenantService *core.SpdTenantSvc) {
+	service = tenantService
+	g := e.Group("/addresses")
+	g.PUT("", handleSetAddresses)
+	g.DELETE("", handleDeleteAddresses)
+	g.GET("", handleGetAddresses)
 }
 
 // handleSetAddresses godoc
 //
-//			@Summary		Update addresses associated with a tenant
-//	 		@Security apiKey
-//		  @Param 			addresses body AddressesMutable true "New addresses to add or change is_signing flag of"
-//			@Produce		json
-//			@Success		200	{object}	ResponseEnvelope{response=Addresses}
-//			@Router			/addresses [put]
+//	@Summary		Update addresses associated with a tenant
+//	@Security		apiKey header string true "Auth token"
+//	@Param			addresses body AddressesMutable true "New addresses to add or change is_signing flag of"
+//	@Produce		json
+//	@Success		200	{object}	ResponseEnvelope{response=Addresses}
+//	@Router			/addresses [put]
 func handleSetAddresses(c echo.Context) error {
-	return c.JSON(http.StatusNotImplemented, map[string]string{})
+	var addresses Addresses
+	err := json.NewDecoder(c.Request().Body).Decode(&addresses)
+
+	if err != nil {
+		return c.JSON(http.StatusInternalServerError, CreateErrorResponseEnvelope(c, http.StatusInternalServerError, err.Error()))
+	}
+
+	var response []string
+	err = service.DB.Transaction(func(tx *gorm.DB) error {
+		for _, address := range addresses {
+			address.TenantID = db.ID(GetTenantContext(c).TenantID)
+
+			res := tx.Model(&address).Updates(&address)
+			if res.Error != nil {
+				return err
+			}
+
+			if res.RowsAffected == 0 {
+				res = tx.Clauses(clause.OnConflict{DoNothing: true}).Create(&address)
+				if res.Error != nil {
+					return err
+				}
+			}
```

Comment:
this is totally my bad - I think we need a separate `POST` endpoint for adding new addreses, separate from the `PUT` which should only update existing ones - so, this `if res.RowsAffected...` block here should not exist (it implies the `PUT` endpoint is an upsert, but I think it's cleaner if we just split out the `CREATE` and `UPDATE` as separate operations i,e POST and PUT )